### PR TITLE
Access out-of-bound checking on CPU backends (Stage 1)

### PR DESCRIPTION
**Pull request by xumingkuan**
Related issue id = #561

Also extends `AssertStmt` from `(message, condition)` to also `(condition, format_str, values)`.

### CODE REVIEW

**Comment by xumingkuan** [comment]
`RuntimeError: [llvm_codegen_utils.cpp:taichi::lang::check_func_call_signature@16] num_params == arglist.size()`

Going to sleep now, will debug it this afternoon.

**Comment by xumingkuan** [comment]
test.py
```python
import taichi as ti

ti.init(print_preprocessed=True, arch=ti.x64, debug=True, print_ir=True)

x = ti.var(ti.i32, shape=(10, 9, 8))

@ti.kernel
def boom():
  x[0, 0, 7] = 1 # OK
  x[0, 0, 8] = 1 # Access out of bound!

boom()
```

**Comment by yuanming-hu** [comment]
> `RuntimeError: [llvm_codegen_utils.cpp:taichi::lang::check_func_call_signature@16] num_params == arglist.size()`
> 
> Going to sleep now, will debug it this afternoon.

It may be caused by the fact that `sprintf` is a function with varargs and the function signature checking does not deal with this case...

**Comment by xumingkuan** [comment]
How to support functions with varargs? Simply removing the function call signature check would result in
```
A s s e r t i o n   f a i l e d :   ( i   > =   F T y - > g e t N u m P a r a m s ( )   | |   F T y - > g e t P a r a m T y p e ( i )   = =   A r g s [ i ] - > g e t T y p e ( ) )   & &   " C a l l i n g   a   f u n c t i o n   w i t h   a   b a d   s i g n a t u r e ! " ,   f i l e   e : \ r e p o s \ l l v m - 8 . 0 . 1 \ l i b \ i r \ i n s t r u c t i o n s . c p p ,   l i n e   3 7 9 
 ```

**Comment by yuanming-hu** [comment]
Could you please print out `type_name(func->getType())` with `func` being the `llvm::Value *` of function `get_assert_msg`?

**Comment by xumingkuan** [comment]
`i8* (i8*, ...)*`

**Comment by yuanming-hu** [comment]
Did you include the format string as an argument to `get_assert_msg`?

File: taichi/codegen/codegen_llvm.h [hunk 1]
```diff
@@ -896,8 +896,18 @@ class CodeGenLLVM : public IRVisitor, public ModuleBuilder {
...
 
   void visit(AssertStmt *stmt) override {
-    stmt->value = call("taichi_assert", get_context(), stmt->val->value,
-                       builder->CreateGlobalStringPtr(stmt->text));
+    Value *text;
+    if (stmt->args.empty()) {
+      text = builder->CreateGlobalStringPtr(stmt->text);
+    } else {
+      std::vector<Value *> args;
+      for (auto arg : stmt->args) {
```
Comment (yuanming-hu) [inline taichi/codegen/codegen_llvm.h]: Maybe we need `text` here in `args`?

Comment (xumingkuan) [reply taichi/codegen/codegen_llvm.h]: Good catch! (but still doesn't work now)

**Comment by xumingkuan** [comment]
```
i8* (i8*, ...)*
[E 03/10/20 15:49:11.787] [codegen_llvm.h:taichi::lang::CodeGenLLVM::finalize_offloaded_task_function@1183] !llvm::verifyFunction(*func, &errs())


Operand is null
  %16 = call i8* (i8*, ...) @get_assert_msg(i8* getelementptr inbounds ([62 x i8], [62 x i8]* @1, i32 0, i32 0), i32 0, i32 0, <null operand!>)
```
Debugging why there is a null operand.

**Comment by xumingkuan** [comment]
Now the assertion works correctly when there are some negative indices (e.g. `x[3, -1, 2] = 1`).
However, indices that exceed the upper bound are replaced by null operands.

**Comment by xumingkuan** [comment]
Oh I found `arg->erased == true` in this case. Does it mean that the `Stmt *` is optimized and thus the `Value` is not calculated?

**Comment by xumingkuan** [comment]
Given that the format string has these `%d`'s, I found it not easy to put sth like `<value optimized out>` at that place...
Why is it optimized out after all?

**Comment by xumingkuan** [comment]
Oh I should have added operands to `AssertStmt`. Fixing.

**Comment by xumingkuan** [comment]
Finally ready for review!

Now I get the following (strange) message when running pbf2d.py. 
```
[E 03/10/20 16:54:23.598] [program.cpp:taichi::lang::assert_failed_host@24] Assertion failure: Accessing Tensor of Size [1200, 100] with indices (8, 74)
```

File: taichi/transforms/ir_printer.cpp [hunk 2]
```diff
@@ -60,7 +60,7 @@ class IRPrinter : public IRVisitor {
   }
 
   void visit(AssertStmt *assert) override {
-    print("{} : assert {}, \"{}\"", assert->id, assert->val->name(),
+    print("{} : assert {}, \"{}\"", assert->id, assert->cond->name(),
           assert->text);
```
Comment (yuanming-hu) [inline taichi/transforms/ir_printer.cpp]: Maybe we should include the arguments here as well?

**Comment by yuanming-hu** [comment]
> Finally ready for review!
> 
> Now I get the following (strange) message when running pbf2d.py.
> 
> ```
> [E 03/10/20 16:54:23.598] [program.cpp:taichi::lang::assert_failed_host@24] Assertion failure: Accessing Tensor of Size [1200, 100] with indices (8, 74)
> ```

Cool! But isn't `(8, 74)` within the tensor range `[1200, 100]`? I agree this looks strange.

File: taichi/runtime/llvm/runtime.cpp [hunk 3]
```diff
@@ -637,6 +638,16 @@ void taichi_assert(Context *context, i32 test, const char *msg) {
   taichi_assert_runtime(context->runtime, test, msg);
 }
 
+const std::size_t ASSERT_MSG_BUFFER_SIZE = 2048;
+char assert_msg_buffer[ASSERT_MSG_BUFFER_SIZE];
+char *get_assert_msg(const char *format, ...) {
+  std::va_list args;
+  va_start(args, format);
+  std::vsnprintf(assert_msg_buffer, ASSERT_MSG_BUFFER_SIZE, format, args);
```
Comment (yuanming-hu) [inline taichi/runtime/llvm/runtime.cpp]: Calling `std::vsnprintf` here may lead to compilation/linking issues when compile/loading the runtime LLVM bitcode. Please consider passing it in as a function pointer, just like `Runtime::host_printf`. You will have to add an argument `Runtime *` to `get_assert_msg`. Use `get_runtime()` to get the pointer to the `LLVMRuntime` struct at codegen. To learn about how to set `LLVMRuntime::host_vsnprintf`, just search for `host_printf` and see how this function is treated in a similar manner.

**Comment by xumingkuan** [comment]
I have `[E 03/10/20 22:04:19.913] Received signal 11 (SIGSEGV)` when running `python test.py` after setting `LLVMRuntime::host_vsnprintf`.

I found that there are `host_printf`'s in both `runtime/runtime.ll` and `llvm/runtime.ll`, but only `host_vsnprintf` in the latter. Is there anywhere I didn't set for `host_vsnprintf`?

**Comment by yuanming-hu** [comment]
Right, you probably need to change here as well: https://github.com/taichi-dev/taichi/blob/master/taichi/program/program.cpp#L159

File: taichi/transforms/ir_printer.cpp (hunk 2 above)
Comment (xumingkuan) [reply taichi/transforms/ir_printer.cpp]: Done. Now it looks like this:
```
14 : assert $13, "Accessing Tensor of Size [1200, 100] with indices (%d, %d)", $1, $2
```

Comment (yuanming-hu) [reply taichi/transforms/ir_printer.cpp]: Awesome!!

**Pull request merged**